### PR TITLE
feat(data type): implement PartialOrd for Variant data-type

### LINKED ISSUE #4903: Column with type: Variant does not support take (while in processor thread 0)
**Summary**

Setup:
```
 create table vartab (id int, v varchar);

insert into vartab (id, v) values 
    (1, '[-1, 12, 289, 2188, false,]'), 
    (2, '{ "x" : "abc", "y" : false, "z": 10} '),
    (3, '{ "bad" : "json", "missing" : true, "close_brace": 10 ');

```

Query:
```
mysql> select id, try_parse_json(v)  from vartab order by id;
ERROR 1105 (HY000): Code: 1010, displayText = Column with type: Variant does not support take (while in processor thread 0).
```

### PR DESCRIPTION
**Pull request by b41sh**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

- introduce `VariantValue` as a wrapper of `serde_json::Value`
- implement `PartialOrd` for `VariantValue`
- rename `JsonColumn` as `VariantColumn`
- rename `DataValue::Json` as `DataValue::Variant`

## Changelog

- New Feature
- Bug Fix

## Related Issues

Fixes #4903

### CODE REVIEW

File: common/datavalues/src/types/variant_value.rs [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+use std::fmt::Display;
+use std::fmt::Formatter;
+use std::ops::Deref;
+
+use common_exception::ErrorCode;
+use common_exception::Result;
+use serde_json::Value;
+
+#[derive(serde::Serialize, serde::Deserialize, Debug, Clone, PartialEq, Eq)]
+pub struct VariantValue(Value);
```
Comment (sundy-li) [inline common/datavalues/src/types/variant_value.rs:26]: Better to put this struct same level as DataValue.

File: common/datavalues/src/types/variant_value.rs [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+            Err(_) => Err(ErrorCode::StrParseError(format!(
+                "Unknown VariantValue: {}",
+                s
+            ))),
+        }
+    }
+}
+
+// VariantValue compares as the following rule:
+// Null > Array > Object > String > Number > Boolean
```
Comment (sundy-li) [inline common/datavalues/src/types/variant_value.rs:75]: So we can give each struct a level value.

If the level is not equal, compare the level value.

File: common/datavalues/src/types/variant_value.rs [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+                        return VariantValue::from(v1).cmp(&VariantValue::from(v2));
+                    }
+                }
+                a1.len().cmp(&a2.len())
+            }
+            (Value::Array(_), _) => Ordering::Greater,
+            (Value::Object(_), Value::Null) => Ordering::Less,
+            (Value::Object(_), Value::Array(_)) => Ordering::Less,
+            (Value::Object(o1), Value::Object(o2)) => {
+                for (k1, k2) in o1.keys().zip(o2.keys()) {
```
Comment (sundy-li) [inline common/datavalues/src/types/variant_value.rs:98]: Is object.keys() sorted ?

Comment (b41sh) [reply common/datavalues/src/types/variant_value.rs:98]: yes, keys are iterated in alphabetical order.

File: common/datavalues/src/types/variant_value.rs (hunk 2 above)
Comment (b41sh) [reply common/datavalues/src/types/variant_value.rs:75]: do you mean add a level filed in `struct VariantValue`?

Comment (sundy-li) [reply common/datavalues/src/types/variant_value.rs:75]: fn level(&self) -> u8 {
  match ...
}

Comment (b41sh) [reply common/datavalues/src/types/variant_value.rs:75]: ok

**Review by sundy-li — APPROVED** [review]

**Review by sundy-li — APPROVED** [review]

**Pull request merged**